gapic-common/lib/gapic/rest/client_stub.rb

```diff
@@ -0,0 +1,143 @@
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require "googleauth"
+
+module Gapic
+  module Rest
+    ##
+    # A class for making REST calls through Faraday
+    # ClientStub's responsibilities:
+    #   - wrap Faraday methods with a bounded explicit interface
+    #   - store service endpoint and create full url for the request
+    #   - store credentials and add auth information to the request
+    #   - convert errors from HTTP responses into ::Gapic::Rest::Error
+    #
+    class ClientStub
+      def initialize endpoint:, credentials:
+        @endpoint = endpoint
+        @endpoint = "https://#{endpoint}" unless /https?:/.match? endpoint
+        @endpoint = @endpoint.chop if @endpoint.to_s.chars.last == "/"
+
+        @credentials = credentials
+      end
+
+      # @param uri [String]
+      # @return [Hash { Symbol => String }]
+      def make_get_request uri:, options:
+        request_url = create_request_url uri
+        headers = {
+          "Authorization" => "Bearer #{@credentials.client.access_token}"
+        }
```

Comment:
We'll need a mechanism for some header cases, including the `x-goog-api-client` header.

Reply:
in progress

Reply:
fixed